### PR TITLE
Handle i18n with redux.

### PR DESCRIPTION
**Pull request by danjm**
@kumavis This is a PR off of your https://github.com/MetaMask/metamask-extension/pull/3572

It can be repointed to master (and probably should be unless you want to keep your PR as the one we review and work off).

One improvement that can be made is to create a new 'translation' component that is connected to state, and pass all translations to that component, instead of modifying the `connect` function. Whether we do that before or after this PR is merged can depend on urgency.

### CODE REVIEW

**Comment by kumavis** [comment]
image
need to also handle all of these

**Comment by kumavis** [comment]
gotta love those PRs that touch every file

File: app/scripts/controllers/preferences.js [hunk 1]
```diff
@@ -11,6 +11,7 @@ class PreferencesController {
       tokens: [],
       useBlockie: false,
       featureFlags: {},
+      currentLocale: 'ja',
```
Comment (kumavis) [inline app/scripts/controllers/preferences.js]: needs to be from `extension.i18n`

**Comment by danjm** [comment]
@kumavis this should be complete now (except for a failing mascara test). You should probably review backend changes, in particular https://github.com/MetaMask/metamask-extension/pull/3609/commits/b9309f689be7f55fde0a32a6e576784aa9fc1061

@tmashuang Is https://github.com/MetaMask/metamask-extension/pull/3609/commits/edf63f8b51ec0717e580ec69b495987a00062b92 okay with you?

@alextsg Front end changes are ready for review.

File: ui/app/components/pending-tx/confirm-send-token.js [hunk 2]
```diff
@@ -22,7 +21,7 @@ const {
 } = require('../../token-util')
 
 const { MIN_GAS_PRICE_HEX } = require('../send/send-constants')
-
+//
```
Comment (alextsg) [inline ui/app/components/pending-tx/confirm-send-token.js]: nit: remove

File: ui/app/components/send/currency-display.js [hunk 3]
```diff
@@ -10,7 +10,7 @@ inherits(CurrencyDisplay, Component)
 function CurrencyDisplay () {
   Component.call(this)
 }
-
+//
```
Comment (alextsg) [inline ui/app/components/send/currency-display.js]: nit: remove

File: ui/app/actions.js [hunk 4]
```diff
@@ -1792,6 +1800,38 @@ function setUseBlockie (val) {
...
+      .then((localeMessages) => {
+        background.setCurrentLocale(key, (err) => {
+          dispatch(actions.hideLoadingIndication())
+          if (err) {
+            return dispatch(actions.displayWarning(err.message))
+          }
+          dispatch(actions.setCurrentLocale(key))
+          dispatch(actions.setLocaleMessages(localeMessages))
+        })
+      })
```
[the remark this answers is not in the document]
Comment (alextsg) [reply ui/app/actions.js]: Just saw that `fetchLocale` always resolves. I guess that works too

**Review by alextsg — APPROVED** [review]

**Comment by kumavis** [comment]
Getting consistent failures in Mascara integration tests

```
  {
    "message": "Uncaught TypeError: Cannot read property '_currentElement' of null\nat dist/mascara/ui.js:135641:7\n\nTypeError: Cannot read property '_currentElement' of null\n    at ReactCompositeComponentWrapper._updateRenderedComponent (dist/mascara/ui.js:102636:53)\n    at ReactCompositeComponentWrapper._performComponentUpdate (dist/mascara/ui.js:102615:10)\n    at ReactCompositeComponentWrapper.updateComponent (dist/mascara/ui.js:102536:12)\n    at ReactCompositeComponentWrapper.receiveComponent (dist/mascara/ui.js:102438:10)\n    at Object.receiveComponent (dist/mascara/ui.js:108495:22)\n    at ReactCompositeComponentWrapper._updateRenderedComponent (dist/mascara/ui.js:102645:23)\n    at ReactCompositeComponentWrapper._performComponentUpdate (dist/mascara/ui.js:102615:10)\n    at ReactCompositeComponentWrapper.updateComponent (dist/mascara/ui.js:102536:12)\n    at ReactCompositeComponentWrapper.receiveComponent (dist/mascara/ui.js:102438:10)\n    at Object.receiveComponent (dist/mascara/ui.js:108495:22)",
```

**Comment by kumavis** [comment]
```
27 03 2018 14:24:21.204:WARN [web-server]: 404: /base/dist/chrome/_locales/undefined/messages.json
```

**Comment by kumavis** [comment]
i notice in the integration tests that even the english locale file fail to be queried

**Comment by kumavis** [comment]
tests passed locally, which is annoying

**Comment by kumavis** [comment]
TESTS ACTUALLY PASS WHAT WHAT

File: app/scripts/lib/get-first-preferred-lang-code.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+const fs = require('fs')
+const path = require('path')
+const extension = require('extensionizer')
+const promisify = require('pify')
+
+const existingLocaleCodes = fs.readdirSync(path.join(__dirname, '..', '..', '_locales'))
```
Comment (kumavis) [inline app/scripts/lib/get-first-preferred-lang-code.js]: we can read `existingLocaleCodes` from the `_locales/index.json`

**Pull request merged**